File: src/main/java/org/vsegda/storage/DataArchiveStorage.java
package org.vsegda.storage;

import com.google.appengine.api.datastore.Cursor;
import org.datanucleus.store.appengine.query.JDOCursorHelper;
import org.vsegda.data.DataArchive;
import org.vsegda.data.DataItem;
import org.vsegda.factory.PM;
import org.vsegda.util.TimeInstant;

import javax.jdo.Query;
import java.util.*;
import java.util.logging.Logger;

/**
 * @author Roman Elizarov
 */
public class DataArchiveStorage {
    private static final Logger log = Logger.getLogger(DataArchiveStorage.class.getName());
    private static final int BATCH_FETCH_SIZE = 10000;

    private DataArchiveStorage() {} // do not create

    public static void storeDataArchive(DataArchive archive) {
        PM.instance().makePersistent(archive);
    }

    @SuppressWarnings({"unchecked"})
    public static List<DataItem> queryItemsFromDataArchives(long streamId, TimeInstant from, TimeInstant to, int nItems) {
        Query query = PM.instance().newQuery(DataArchive.class);
        String queryFilter = "streamId == _id";
        String queryParams = "long _id";
        Map<String, Object> queryArgs = new HashMap<String, Object>();
        queryArgs.put("_id", streamId);
        if (from != null) {
            queryFilter += " && firstTimeMillis >= _from";
            queryParams += ", long _from";
            queryArgs.put("_from", from.time());
        }
        if (to != null) {
            queryFilter += " && firstTimeMillis < _to";
            queryParams += ", long _to";
            queryArgs.put("_to", to.time());
        }
        query.setFilter(queryFilter);
        query.declareParameters(queryParams);
        query.setOrdering("firstTimeMillis desc");
        int range = 1 + nItems / DataArchive.COUNT_ESTIMATE; // estimate number of archives
        query.setRange(0, range);

        List<DataItem> items = new ArrayList<DataItem>();
        long queryStartTime = System.currentTimeMillis();
        int totalArchives = 0;

    loadLoop:
        while (true) {
            List<DataArchive> archives = new ArrayList<DataArchive>((Collection<DataArchive>)query.executeWithMap(queryArgs));
            Cursor cursor = JDOCursorHelper.getCursor(archives);
            totalArchives += archives.size();

            for (DataArchive archive : archives) {
                int prevSize = items.size();
                items.addAll(archive.getItems());
                Collections.reverse(items.subList(prevSize, items.size())); // turn ascending items into descending
                if (items.size() >= nItems) {
                    items.subList(nItems, items.size()).clear();
                    break loadLoop;
                }
            }
            if (archives.size() < range)
                break; // returns less data, means no stuff -- done

            // load next batch using JDO cursor
            Map<String, Object> extensionMap = new HashMap<String, Object>();
            extensionMap.put(JDOCursorHelper.CURSOR_EXTENSION, cursor);
            query.setExtensions(extensionMap);
        }

        log.info(String.format("(streamId=%d, from=%s, to=%s, nItems=%d) retrieved %d data archives from storage with %d items in %d ms",
                streamId, from, to, nItems, totalArchives, items.size(), System.currentTimeMillis() - queryStartTime));
        Collections.reverse(items); // turn descending into ascending order
        return items;
    }

    @SuppressWarnings("unchecked")
    public static void updateStreamId(long fromId, long toId) {
        Query query = PM.instance().newQuery(DataArchive.class);
        query.setFilter("streamId == _id");
        query.declareParameters("long _id");
        query.getFetchPlan().setFetchSize(BATCH_FETCH_SIZE);
        Collection<DataArchive> items = (Collection<DataArchive>) query.execute(fromId);
        for (DataArchive item : items)
            item.setStreamId(toId);
    }

    @SuppressWarnings("unchecked")
    public static void removeAllByStreamId(long streamId) {
        Query query = PM.instance().newQuery(DataArchive.class);
        query.setFilter("streamId == _id");
        query.declareParameters("long _id");
        query.getFetchPlan().setFetchSize(BATCH_FETCH_SIZE);
        Collection<DataArchive> items = (Collection<DataArchive>) query.execute(streamId);
        for (DataArchive item : items)
            PM.instance().deletePersistent(item);
    }
}

File: service/src/main/java/org/vsegda/storage/DataArchiveStorage.kt
package org.vsegda.storage

import com.google.appengine.api.datastore.*
import org.vsegda.data.*
import org.vsegda.util.*
import java.util.*

private var Entity.streamId by Prop.long
private var Entity.count by Prop.int
private var Entity.firstValue by Prop.double
private var Entity.firstTimeMillis by Prop.long
private var Entity.highValue by Prop.double
private var Entity.lowValue by Prop.double
private var Entity.encodedItems by Prop.blob

object DataArchiveStorage : BaseStorage<DataArchive>() {
    override val kind: String = "DataArchive"

    private fun keyOf(archiveId: Long) = KeyFactory.createKey(kind, archiveId)

    override fun DataArchive.toKey() =
        if (archiveId == 0L) null else keyOf(archiveId)

    override fun DataArchive.toEntity() = newEntity { e ->
        e.streamId = streamId
        e.count = count
        e.firstValue = firstValue
        e.firstTimeMillis = firstTimeMillis
        e.highValue = highValue
        e.lowValue = lowValue
        e.encodedItems = encodedItems
    }

    override fun Entity.toObject() = DataArchive().apply {
        val e = this@toObject
        archiveId = e.key?.id ?: 0L
        streamId = e.streamId ?: 0L
        count = e.count ?: 0
        firstValue = e.firstValue ?: Double.NaN
        firstTimeMillis = e.firstTimeMillis ?: 0L
        highValue = e.highValue ?: Double.NaN
        lowValue = e.lowValue ?: Double.NaN
        encodedItems = e.encodedItems
    }

    fun storeDataArchive(archive: DataArchive) =
        logged({ "storeDataArchive($archive) -> $it" }) {
            store(archive).also { archive.archiveId = it.id }
        }

    fun queryItemsFromDataArchives(streamId: Long, from: TimeInstant?, to: TimeInstant?, nItems: Int): List<DataItem> =
        logged("queryItemsFromDataArchives(streamId=$streamId, from=$from, to=$to, nItems=$nItems)",
            result={ "${it.size} items" }, around=true
        ) {
            val estimatedRange = 1 + nItems / ARCHIVE_COUNT_ESTIMATE // estimate number of archives
            val items = ArrayList<DataItem>()
            val it = query {
                filterEq(Entity::streamId, streamId)
                if (from != null) filterGreaterEq(Entity::firstTimeMillis, from.time())
                if (to != null) filterLess(Entity::firstTimeMillis, to.time())
                sortDescBy(Entity::firstTimeMillis)
            }.asSequence(estimatedRange)
            for (archive in it) {
                val archiveItems = archive.items
                log.info("Retrieved archive $archive, decoded ${archiveItems.size} items")
                val prevSize = items.size
                items.addAll(archiveItems)
                items.subList(prevSize, items.size).reverse() // turn ascending items into descending
                if (items.size >= nItems) {
                    items.subList(nItems, items.size).clear()
                    break
                }
            }
            items.reverse() // turn descending into ascending order
            items
        }

    fun updateStreamId(fromId: Long, toId: Long) =
        logged("updateStreamId(fromId=$fromId, toId=$toId)") {
            queryAll(fromId).onEach { it.streamId = toId }.store()
        }

    fun removeAllByStreamId(streamId: Long) =
        logged("removeAllByStreamId($streamId)") {
            queryAll(streamId).map { it.key }.delete()
        }

    private fun queryAll(streamId: Long): Sequence<Entity> =
        query { filterEq(Entity::streamId, streamId) }.asEntitySequence()
}
